i18n の課題

DeeeU opened the pull request:
## 条件
- 英語と日本語の翻訳をconfig/locales配下のYAMLファイルに定義する
  - [x] i18nの導入

- 各画面の表示を翻訳ファイルから取得して表示できるようにする
  - [x] ja.ymlの導入
　
- デフォルトの表示言語を英語または日本語に切り替えることで、実際に画面の表示が英語と日本語に切り替わる（要スクリーンショット）
  - [x] i18nによるホーム画面の日本語化
  - [x] i18nによる詳細画面の日本語化
  - [x] i18nによる新規作成画面の日本語化
  - [x] i18nによる編集画面の日本語化
  - [x] i18nによる削除時のメッセージの日本語化

- 表示を切り替えるときは「アプリケーションの設定ファイルを変更してサーバー再起動」でOK（ただし、提出時は日本語をデフォルトにすること）
  - [x] application.rbへの設定記載

- [Rails 国際化 (i18n) API - Railsガイド](https://railsguides.jp/i18n.html) を読み、使えそうなメソッドがあれば積極的に使う

- rubocopをパスさせる（要スクリーンショット）
  - [x] rubocopで確認
----------------------------------------------------------------------------------------------------------------------------

mpg-yutaro-kanabuchi committed:
[fix] gemの追加

mpg-yutaro-kanabuchi committed:
[add] 設定及び必要ファイルの追加

mpg-yutaro-kanabuchi committed:
[fix]conflict

mpg-yutaro-kanabuchi committed:
[fix]不要部分の削除

mpg-yutaro-kanabuchi committed:
[fix]不要部分の削除

mpg-yutaro-kanabuchi committed:
[fix] ホーム画面の日本語化

mpg-yutaro-kanabuchi committed:
[fix] 詳細画面の日本語化

mpg-yutaro-kanabuchi committed:
[fix] 新規作成画面の日本語化

mpg-yutaro-kanabuchi committed:
[fix] 削除及び更新時のメッセージの日本語化

mpg-yutaro-kanabuchi committed:
[fix] 編集画面の日本語化

mpg-yutaro-kanabuchi committed:
[fix] rubocopによる書式の変更

JunichiIto commented on app/controllers/books_controller.rb:
```diff
@@ -28,7 +28,7 @@ def create
     respond_to do |format|
       if @book.save
-        format.html { redirect_to @book, notice: 'Book was successfully created.' }
+        format.html { redirect_to @book, notice: (t 'message.create_info') }
```
```suggestion
        format.html { redirect_to @book, notice: t('message.create_info') }
```

というスタイルで書くことの方が多いと思いますー。

JunichiIto commented on app/views/books/index.html.erb:
```diff
@@ -5,10 +5,10 @@
-      <th>Memo</th>
-      <th>Author</th>
-      <th>Picture</th>
+      <th><%= t  'book_info.title' %></th>
```
```suggestion
      <th><%= t 'book_info.title' %></th>
```

スペースがひとつ多いですね。

あと、モデルの属性名は`human_attribute_name`を使うとRailsらしくなります。

DeeeU replied:
> あと、モデルの属性名はhuman_attribute_nameを使うとRailsらしくなります。

この記述に関してドキュメントを読んで実行しようとしましたが自分ではNameErrorが出てしまい上手く参照出来なかったので変更出来ませんでした。
[参考にしたドキュメントの記事](https://railsguides.jp/i18n.html#:~:text=4.5%20Active%20Record%E3%83%A2%E3%83%87%E3%83%AB%E3%81%A7%E7%BF%BB%E8%A8%B3%E3%82%92%E8%A1%8C%E3%81%AA%E3%81%86)
Qiitaなども参考にしましたがどこも設定ファイルの設定方法の記載がないのでja.yamlと書式の表示を変えたいファイル（今回だと.erb）のみを変更するという認識です。
参考記事 : https://fuqda.hatenablog.com/entry/2019/04/07/212254
参考記事内の ` t("activerecord.attributes.task.name")`は上手く参照してくれるのですが `Task.human_attribute_name(:name)` だとTaskでName Errorが出ます。
これを通すにはTaskのmodelを作るのでしょうか...

JunichiIto replied:
`Task.human_attribute_name(:name)`で冒頭に`Task`と書いているのは、記事で使われているRailsアプリケーションにTaskモデルが存在しているからですね。
一方、このRailsアプリにはTaskクラスはありません。

つまり、ここに書くべきクラス名は・・・？

てな感じでヒントになるでしょうか？（クラス名だけでなく属性名にも注意してくださいね）

JunichiIto commented on app/views/books/index.html.erb:
```diff
@@ -20,14 +20,14 @@
-        <td><%= link_to 'Destroy', book, method: :delete, data: { confirm: 'Are you sure?' } %></td>
+        <td><%= link_to (t 'button.show'), book %></td>
+        <td><%= link_to (t 'button.edit'), edit_book_path(book) %></td>
+        <td><%= link_to (t 'button.destroy'), book, method: :delete, data: { confirm: (t 'message.finalconfirmation') } %></td>
```
`finalconfirmation`という単語はないので`final_confirmation`のように区切った方がいいですね。
ただ、finalはちょっと強すぎるので`confirmation`とか`confirm`だけでもいい気がします。

JunichiIto commented on config/locales/ja.yml:
```diff
@@ -0,0 +1,22 @@
+book_info: &book_info
```
`&book_info`みたいなテクニックは特別な理由がない限り使わない方がシンプルで管理しやすいと思いますー。

JunichiIto reviewed (commented):
コメントしました。あと、以下の点も確認をお願いします。

- h1タグのBooksやNew Bookのような見出しも翻訳してください
- 日本語表示だけでなく、英語で表示したときのスクショも載せてください

mpg-yutaro-kanabuchi committed:
Change: 変数名finalconfirmationからconfirmへの変更
finalは意味が強く、今回の確認メッセージに対しては不適合な為変更

mpg-yutaro-kanabuchi committed:
-F-

mpg-yutaro-kanabuchi committed:
add: ページごとのタイトル名の追記
ページごとのタイトルが翻訳されない状態であったのでpata_titleに各ページのタイトルを追記した

mpg-yutaro-kanabuchi committed:
change: 各ページのタイトル名変更
タイトル名が固定であり日本語に翻訳されていなかった為、ja.ymlに対応したタイトルを記述した

DeeeU commented:
# 指摘箇所のリスト及びやることリスト
## 指摘箇所
  https://github.com/DeeeU/fjord-books_app/pull/1#pullrequestreview-1037095910

## やることリスト
  - [ ] ページのタイトルの翻訳
  - [add: ページごとのタイトル名の追記](https://github.com/DeeeU/fjord-books_app/pull/1/commits/8d946e44d18201247d1c1bac8853bece6a793aad)
  - [change: 各ページのタイトル名変更](https://github.com/DeeeU/fjord-books_app/pull/1/commits/48fdd978a98af7673ec82a7a6db4410e15d952f5)
  
  - [ ] 英語表示でのスクショの添付
    - [ ] index
    - [ ] new
    - [ ] show
    - [ ] detail
  　
  - [ ] 書式の変更
  https://github.com/DeeeU/fjord-books_app/pull/1#discussion_r919888625
  - [change: 書式の変更　(t,) ==> t()](https://github.com/DeeeU/fjord-books_app/pull/1/commits/7669accdf6ee701526e166f29ffc3d7badd76cc6)
  
  - [ ] 変数名の変更
  https://github.com/DeeeU/fjord-books_app/pull/1#discussion_r919890963
  - [Change: 変数名finalconfirmationからconfirmへの変更](https://github.com/DeeeU/fjord-books_app/pull/1/commits/7bacab1ab1234e1c187dee053bb3457991248c43) 
  
  - [ ] 不要なスペースの削除
  https://github.com/DeeeU/fjord-books_app/pull/1#discussion_r919889166
  - [スペースの削除](https://github.com/DeeeU/fjord-books_app/pull/1/commits/f687a9f3a5be753973db3274f1a44e0d17232fe1)

mpg-yutaro-kanabuchi committed:
change: 書式の変更　(t,) ==> t()
あまり見られない書き方であるために今後コードを見る際に可読性を低くする可能性がある。
ドキュメントでは変更後の書式を用いているためドキュメントで記載されている書式に変更
https://railsguides.jp/i18n.html#:~:text=%3D%20%3Ade-,4.4%20%E5%AE%89%E5%85%A8%E3%81%AAHTML%E5%A4%89%E6%8F%9B,-%E3%82%AD%E3%83%BC%E5%90%8D%E3%82%92

mpg-yutaro-kanabuchi committed:
change: label書式の変更

mpg-yutaro-kanabuchi committed:
change: 書式変更

DeeeU committed:
Merge branch '02-i18n' into my-i18n

JunichiIto commented on Gemfile:
```diff
@@ -59,3 +59,5 @@ gem 'net-smtp'
 gem 'carrierwave'
 gem 'rails-i18n'
+
+gem 'kaminari'
```
もしかして次のプラクティスの変更が混じってます？
こういううっかりを避けるためにコミットする前にセルフレビューする癖を付けましょう。

https://blog.jnito.com/entry/2021/01/30/172607

JunichiIto commented on config/locales/ja.yml:
```diff
@@ -1,22 +1,24 @@
-  picture: "写真"
+ja:
+  page_title:
+    index: "本"
```
これだと、新しいページ（たとえば日報一覧）を作ったときに`page_title.index`が使えなくなりますね（日報一覧なのに「本」と表示されるので）。
本のページでしか使えない翻訳にはkeyに`book`を含めて他と被らないようにした方が、のちのち便利です。

ちなみに遅延探索のテクニックを使うと他のページと被りにくいキーを付けやすくなります。
https://railsguides.jp/i18n.html#%E9%81%85%E5%BB%B6%E6%8E%A2%E7%B4%A2

JunichiIto commented on config/locales/ja.yml:
```diff
@@ -1,22 +1,24 @@
+    show: "表示"
+    edit: "編集"
+    destroy: "削除"
+    newbook: "新しい本の追加"
```
```suggestion
    new: "新しい%{name}の追加"
```

のように翻訳の一部をパラメータ化すると翻訳を再利用しやすくなります。

JunichiIto commented on config/locales/ja.yml:
```diff
@@ -1,22 +1,24 @@
+
+  message:
+    confirm: "本当によろしいですか？"
+    create_info: "本の情報は作成されました"
```
これも再利用性の観点でいうと「本」をパラメータ化した方が良さそうです。

JunichiIto reviewed (commented):
コメントしました！

DeeeU committed:
Merge branch '02-i18n' into my-i18n

mpg-yutaro-kanabuchi committed:
change: human_attribute_nameの導入とviewファイルでの参照変更
 Githubでのコメントでhuman_attribute_nameによってrailsらしくなるとのコメントがあり使ってみよう！との思いで導入をした
今回気にしている点はhuman_attribute_nameでのattributesの下で定義しているbookが今後の遅延探索で記述するbooksと類似しているため可読性を低くしてしまうのではないか？という点。また、今回のattributesの下のbookを別の名称にすればいいと思うが、他の名称にすることで可読性が低くなる恐れがありそう。
今回はそれらの点を含ませながらpushした

mpg-yutaro-kanabuchi committed:
change: パラメータ化及びパラメータ化に伴う書式の変更
翻訳の一部をパラメータ化することによって再利用性が高まるので今回はパラメータ化が出来た部分の変更とviewファイル,controllerの一部を変更した。

mpg-yutaro-kanabuchi committed:
change: 遅延探索によるタイトルの翻訳の導入
前回までの翻訳の定義だとページを追加する際に再定義が必要となるので遅延探索によってviewファイルごとで定義し再定義を行わずにするようにした

mpg-yutaro-kanabuchi committed:
change: rubocopによる指摘点変更

JunichiIto commented on app/views/books/_form.html.erb:
```diff
@@ -12,22 +12,22 @@
 
   <div class="field">
-    <%= form.label t('book_info.title') %>
+    <%= form.label 'title' %>
```
`form.text_field :title`と同じようにシンボルを渡す方が一貫性が出そうです

JunichiIto commented on app/views/books/index.html.erb:
```diff
@@ -30,5 +30,4 @@
 
-<%= link_to t('button.newbook'), new_book_path %>
-<%= paginate @books %>
+<%= link_to t('button.new', name: "本"), new_book_path %>
```
`"本"`の部分もi18n化しないと、英語表示でも「本」が出てきそうです

JunichiIto commented on app/views/books/index.html.erb:
```diff
@@ -1,14 +1,14 @@
-      <th><%= t('book_info.memo') %></th>
-      <th><%= t('book_info.author') %></th>
-      <th><%= t('book_info.picture') %></th>
+      <th><%= Book.human_attribute_name("title") %></th>
```
ここもシンボルを渡すようにしましょう

JunichiIto commented on config/application.rb:
```diff
@@ -22,6 +22,6 @@ class Application < Rails::Application
 
     # デフォルトの設定(日本語へ)
-    I18n.default_locale = :ja
+    I18n.default_locale = :en
```
コメントと実装が一致しないですね。
そもそもコメントがいらないのでは？

[【動画付き・初心者向け】コメントにはWHYを書こう 〜不要なコメントと必要なコメントの違いについて〜 \- Qiita](https://qiita.com/jnchito/items/f0d90af4ed44b7484103)

JunichiIto commented on config/locales/ja.yml:
```diff
@@ -1,24 +1,31 @@
-    destroy_info: "本の情報は削除されました"
+    create_info: "「 %{name} 」の情報は作成されました"
+    update_info: "「 %{name} 」の情報は更新されました"
+    destroy_info: "「 %{name} 」の情報は削除されました"
```
日本語で追加した翻訳は英語にも追加するようにしてください〜。

JunichiIto reviewed (commented):
コメントしました！

mpg-yutaro-kanabuchi committed:
change: デフォルトを日本語化

mpg-yutaro-kanabuchi committed:
change: シンボルによる受け渡し

mpg-yutaro-kanabuchi committed:
change: 表示していた本の翻訳

mpg-yutaro-kanabuchi committed:
change: 英語での翻訳追加及び日本語の翻訳追加(items)

JunichiIto commented on config/locales/en.yml:
```diff
@@ -46,3 +16,19 @@ en:
+
+  message:
+    confirm: "Are you ok ?"
+    create_info: "「 %{name} 」 was created"
```
細かいですが、カギ括弧は日本人向けのような気が・・・。
そもそも、日本語の翻訳も含めてカギ括弧はなしでもいい気がしますね。

JunichiIto replied:
あれ、こっちのカギ括弧はなくさないんですか？

JunichiIto commented on app/views/books/index.html.erb:
```diff
@@ -30,4 +30,4 @@
 <br>
 
-<%= link_to t('button.new', name: "本"), new_book_path %>
+<%= link_to t('button.new', name: t('items.book')), new_book_path %>
```
https://railsguides.jp/i18n.html#active-record%E3%83%A2%E3%83%87%E3%83%AB%E3%81%A7%E7%BF%BB%E8%A8%B3%E3%82%92%E8%A1%8C%E3%81%AA%E3%81%86 の`Model.model_name.human`を使った方がRailsらしくなります〜。

JunichiIto commented on app/views/books/edit.html.erb:
```diff
@@ -1,6 +1,6 @@
 
-<%= link_to 'Show', @book %> |
-<%= link_to 'Back', books_path %>
+<%= link_to t('button.show'), @book %> |
```
細かいことをいうと、ここはボタンではなくリンクですね。
あまりUIに依存しないkey名を付けた方が再利用性が高くなりそうです。

JunichiIto commented on config/application.rb:
```diff
@@ -18,5 +18,10 @@ class Application < Rails::Application
+    I18n.load_path += Dir[Rails.root.join('lib', 'locale', '*.{rb, yml}')]
+    I18n.available_locales = [:en, :ja]
+
+    # デフォルトの設定(日本語へ)
```
このコメントはWHYを書いてるわけではないので、消してしまいましょう。

JunichiIto reviewed (commented):
コメントしました〜。

DeeeU committed:
change: コメントの削除

DeeeU committed:
change: Model.model_name.humanの導入

DeeeU committed:
change: カギ括弧の削除

DeeeU committed:
change: buttonからitemsへの変更

JunichiIto reviewed (commented):
1点コメントしました〜。

DeeeU committed:
change: 英語翻訳内のカギ括弧の削除

JunichiIto reviewed (approved):
確認しました。OKです 🙆‍♂️
